lib/api/api_auth_webauthn.go

```diff
@@ -0,0 +1,476 @@
+		}
+
+		if err := unmarshalTo(r.Body, &req); err != nil {
+			l.Debugln("Failed to parse response:", err)
+			http.Error(w, "Failed to parse response.", http.StatusBadRequest)
+			return
+		}
+
+		parsedResponse, err := req.Credential.Parse()
+		if err != nil {
+			l.Debugln("Failed to parse WebAuthn authentication response", err)
+			badRequest(w)
+			return
+		}
+
+		updatedCred, err := s.engine.ValidateLogin(s.user(guiCfg), state, parsedResponse)
+		if err != nil {
+			l.Infoln("WebAuthn authentication failed", err)
+
+			if state.UserVerification == webauthnProtocol.VerificationRequired {
+				antiBruteForceSleep()
+				http.Error(w, "Conflict", http.StatusConflict)
+				return
+			}
+
+			forbidden(w)
+			return
+		}
+
+		authenticatedCredId := base64.URLEncoding.EncodeToString(updatedCred.ID)
+
+		persistentState, err := s.loadState()
+		if err != nil {
+			l.Warnln("Failed to load persistent WebAuthn state", err)
+			http.Error(w, "Internal server error", http.StatusInternalServerError)
+			return
+		}
+
+		for _, cred := range persistentState.Credentials {
+			if cred.ID == authenticatedCredId {
+				if cred.RequireUv && !updatedCred.Flags.UserVerified {
+					antiBruteForceSleep()
+					http.Error(w, "Conflict", http.StatusConflict)
+					return
+				}
+				break
+			}
+		}
+
+		authenticatedCredName := authenticatedCredId
+		var signCountBefore uint32 = 0
+
+		updateCredIndex := slices.IndexFunc(persistentState.Credentials, func(cred WebauthnCredential) bool { return cred.ID == authenticatedCredId })
+		if updateCredIndex != -1 {
+			updateCred := &persistentState.Credentials[updateCredIndex]
+			signCountBefore = updateCred.SignCount
+			authenticatedCredName = updateCred.NicknameOrID()
+			updateCred.SignCount = updatedCred.Authenticator.SignCount
+			updateCred.LastUseTime = time.Now().Truncate(time.Second).UTC()
+			err = s.storeState(persistentState)
+			if err != nil {
+				l.Warnln("Failed to update authenticated WebAuthn credential", err)
+				http.Error(w, "Internal server error", http.StatusInternalServerError)
+				return
+			}
+		}
+
+		if updatedCred.Authenticator.CloneWarning && signCountBefore != 0 {
+			l.Warnln(fmt.Sprintf("Invalid WebAuthn signature count for credential %q: expected > %d, was: %d. The credential may have been cloned.", authenticatedCredName, signCountBefore, parsedResponse.Response.AuthenticatorData.Counter))
+		}
+
+		tokenCookieManager.createSession(guiCfg.User, req.StayLoggedIn, w, r)
+		w.WriteHeader(http.StatusNoContent)
+	}
+}
+
+func (s *webauthnService) getConfigLikeState(w http.ResponseWriter, _ *http.Request) {
+	persistentState, err := s.loadState()
+	if err != nil {
+		l.Warnln("Failed to load persistent WebAuthn state", err)
+		http.Error(w, "Internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	sendJSON(w, persistentState)
+}
+
+func (s *webauthnService) updateConfigLikeState(w http.ResponseWriter, r *http.Request) {
```

Comment:
Similarly for these things that are read-modify-write without any locking or session handling. Should there be some exclusivity here that only a single client is doing this at the time?

Reply:
Yeah, good point. This is a bit more impactful than https://github.com/syncthing/syncthing/pull/9175#discussion_r1730864774 as this could end up causing data loss, as opposed to just failing an authentication ceremony and just trying again.

I'm not sure a (server-side-only) locking mechanism would really work here with the way this works with the settings GUI (see also https://github.com/emlun/syncthing/pull/7), as the GUI submits the whole state at once rather than updating one property at a time. You'd have to extend the lock to the client side too, acquiring it when the user starts editing config and releasing it when the user submits or cancels the changes (or after a timeout, if the user abandons the settings GUI without closing it).

Alternatively, one solution could be that before saving the state to config, the server generates and sets a "version tag" on the config object, and when updating the config, the server only accepts the request if the incoming request has the same version tag as the one currently stored (and if not, the server responds `409 CONFLICT` or similar).

Though I guess that could be a poor user experience if you click "Save" in the settings GUI, get an error and the instruction to reload and try again, and then have to redo all your config changes from scratch. On the other hand there's no locking or conflict detection like that for the rest of the settings GUI at the moment.

I don't have strong opinions against or in favour of any of these. What do you think?

Reply:
I'm surprised there isn't an established recipe of some kind for these kinds of registrations?

Reply:
If you mean an established recipe for this state handling in WebAuthn applications generally, I think the general pattern would be to use database transactions to prevent data races and the like.

Anyway, is this still relevant now with https://github.com/emlun/syncthing/pull/7 merged? As discussed there (https://github.com/emlun/syncthing/pull/7#issuecomment-2312254000 onwards), the state that changes on each login is now kept in the DB (this is the `WebauthnVolatileState` struct) apart from the state kept in config which changes only when actively edited by the user (this is the `GUIConfiguration.WebauthnState` struct). `WebauthnVolatileState` is mostly informational and it's no big deal if it gets overwritten by a data race or erased by the user resetting the DB. This is also documented in the protobuf definition:

https://github.com/syncthing/syncthing/blob/bc35c942c775f88649237a19a45a09c1c6b0180f/proto/lib/api/api_auth_webauthn.proto#L7-L9

Does that suffice, or do we need to do more here?